server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java

```diff
@@ -302,43 +319,149 @@ private DirectDownloadCommand getDirectDownloadCommandFromProtocol(DownloadProto
+        String alias = certificateVO.getAlias();
+
+        s_logger.debug("Uploading certificate: " + certificateId + "(" + alias + ") to host " + hostId);
+
+        SetupDirectDownloadCertificate cmd = new SetupDirectDownloadCertificate(certificate, alias);
         Answer answer = agentManager.easySend(hostId, cmd);
         if (answer == null || !answer.getResult()) {
+            String msg = "Certificate " + alias + " could not be added to host " + hostId;
+            if (answer != null) {
+                msg += " due to: " + answer.getDetails();
+            }
+            s_logger.debug(msg);
             return false;
         }
-        s_logger.info("Certificate " + certificateName + " successfully uploaded to host: " + hostId);
+
+        DirectDownloadCertificateHostMapVO mapVO = new DirectDownloadCertificateHostMapVO(certificateId, hostId);
+        directDownloadCertificateHostMapDao.persist(mapVO);
+
+        s_logger.info("Certificate " + alias + " successfully uploaded to host: " + hostId);
         return true;
     }
+
+    @Override
+    public boolean configure(String name, Map<String, Object> params) throws ConfigurationException {
+        backgroundPollManager.submitTask(new DirectDownloadCertificateUploadBackgroundTask(this,
+                hostDao, dataCenterDao, directDownloadCertificateDao, directDownloadCertificateHostMapDao));
+        return true;
+    }
+
+    @Override
+    public String getConfigComponentName() {
+        return DirectDownloadManager.class.getSimpleName();
+    }
+
+    @Override
+    public ConfigKey<?>[] getConfigKeys() {
+        return new ConfigKey<?>[]{
+                DirectDownloadCertificateUploadInterval
+        };
+    }
+
+    public static final class DirectDownloadCertificateUploadBackgroundTask extends ManagedContextRunnable implements BackgroundPollTask {
+        private DirectDownloadManager directDownloadManager;
+        private HostDao hostDao;
+        private DirectDownloadCertificateDao directDownloadCertificateDao;
+        private DirectDownloadCertificateHostMapDao directDownloadCertificateHostMapDao;
+        private DataCenterDao dataCenterDao;
+
+        public DirectDownloadCertificateUploadBackgroundTask(
+                final DirectDownloadManager caManager,
+                final HostDao hostDao,
+                final DataCenterDao dataCenterDao,
+                final DirectDownloadCertificateDao directDownloadCertificateDao,
+                final DirectDownloadCertificateHostMapDao directDownloadCertificateHostMapDao) {
+            this.directDownloadManager = caManager;
+            this.hostDao = hostDao;
+            this.dataCenterDao = dataCenterDao;
+            this.directDownloadCertificateDao = directDownloadCertificateDao;
+            this.directDownloadCertificateHostMapDao = directDownloadCertificateHostMapDao;
+        }
+
+        @Override
+        protected void runInContext() {
```

Comment:
I am confused here. Is this "uploadCertificate" command being constantly executed?
I mean, it is an async task. In ACS the operator/admin will register a certificate to be used in hosts, and then this thread will pick up this request (in async mode) and process it?